sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c

```diff
@@ -0,0 +1,387 @@
+            /* exact match */
+            DEBUG("  %p is an exact match\n", (void *)tmp);
+            return tmp;
+        }
+        if ((node == NULL) && (tmp->mode == _EMPTY)) {
+            node = tmp;
+        }
+    }
+    if (node != NULL) {
+        DEBUG("  using %p\n", (void *)node);
+        _override_node(addr, iface, node);
+    }
+#if ENABLE_DEBUG
+    else {
+        DEBUG("  NIB full\n");
+    }
+#endif
+    return node;
+}
+
+static inline bool _is_gc(_nib_onl_entry_t *node)
+{
+    return ((node->mode & ~(_NC)) == 0) &&
+           ((node->info & GNRC_IPV6_NIB_NC_INFO_AR_STATE_MASK) ==
+            GNRC_IPV6_NIB_NC_INFO_AR_STATE_GC);
+}
+
+static inline _nib_onl_entry_t *_cache_out_onl_entry(const ipv6_addr_t *addr,
+                                                     unsigned iface,
+                                                     uint16_t cstate)
+{
+    /* Use clist as FIFO for caching */
+    _nib_onl_entry_t *first = (_nib_onl_entry_t *)clist_lpop(&_next_removable);
+    _nib_onl_entry_t *tmp = first, *res = NULL;
+
+    DEBUG("nib: Searching for replaceable entries (addr = %s, iface = %u)\n",
+          ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+    if (tmp != NULL) {
+        do {
+            if (_is_gc(tmp)) {
+                DEBUG("nib: Removing neighbor cache entry (addr = %s, "
+                      "iface = %u) ",
+                      ipv6_addr_to_str(addr_str, &tmp->ipv6,
+                                       sizeof(addr_str)),
+                      _nib_onl_get_if(tmp));
+                DEBUG("for (addr = %s, iface = %u)\n",
+                      ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)),
+                      iface);
+                res = tmp;
+                res->mode = _EMPTY;
+                _override_node(addr, iface, res);
+                /* masked above already */
+                res->info = cstate;
+                res->mode = _NC;
+            }
+            else {
+                /* requeue if not garbage collectible at the moment */
+                DEBUG("nib: Requeing (addr = %s, iface = %u)\n",
+                      ipv6_addr_to_str(addr_str, &tmp->ipv6,
+                                       sizeof(addr_str)),
+                      _nib_onl_get_if(tmp));
+                clist_rpush(&_next_removable, (clist_node_t *)tmp);
+                tmp = (_nib_onl_entry_t *)clist_lpop(&_next_removable);
+            }
+        } while (tmp != first);
+    }
+    return res;
+}
+
+_nib_onl_entry_t *_nib_nc_add(const ipv6_addr_t *addr, unsigned iface,
+                              uint16_t cstate)
+{
+    cstate &= GNRC_IPV6_NIB_NC_INFO_NUD_STATE_MASK;
+    assert(cstate != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_DELAY);
+    assert(cstate != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_PROBE);
+    assert(cstate != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_REACHABLE);
+    _nib_onl_entry_t *node = _nib_onl_alloc(addr, iface);
+    if (node != NULL) {
```

Comment:
I would prefer
```c
if (node == NULL) {
    return NULL;
}
```
to remove the indentation below. Leads to a much better readability of the code.